62.CAD/main.cpp

```diff
@@ -48,21 +47,160 @@ struct float4
 
 typedef nbl::core::vector2d<double> double2;
 typedef nbl::core::vector2d<uint32_t> uint2;
+typedef uint32_t uint;
+
+// TODO: Use a math lib?
+double dot(const double2& a, const double2& b)
+{
+	return a.X * b.X + a.Y * b.Y;
+}
 
 #include "common.hlsl"
 
+bool operator==(const LineStyle& lhs, const LineStyle& rhs)
+{
+	const bool areParametersEqual =
+		lhs.color == rhs.color &&
+		lhs.screenSpaceLineWidth == rhs.screenSpaceLineWidth &&
+		lhs.worldSpaceLineWidth == rhs.worldSpaceLineWidth &&
+		lhs.stipplePatternSize == rhs.stipplePatternSize &&
+		lhs.recpiprocalStipplePatternLen == rhs.recpiprocalStipplePatternLen &&
+		lhs.phaseShift == rhs.phaseShift;
+
+	if (!areParametersEqual)
+		return false;
+
+	if (lhs.stipplePatternSize == -1)
+		return true;
+
+	assert(lhs.stipplePatternSize <= LineStyle::STIPPLE_PATTERN_MAX_SZ && lhs.stipplePatternSize >= 0);
+
+	const bool isStipplePatternArrayEqual = std::memcmp(lhs.stipplePattern, rhs.stipplePattern, sizeof(decltype(lhs.stipplePatternSize)) * lhs.stipplePatternSize);
+
+	return areParametersEqual && isStipplePatternArrayEqual;
+}
+
+// holds values for `LineStyle` struct and caculates stipple pattern re values, cant think of better name
+class CPULineStyle
+{
+public:
+	static const uint32_t STIPPLE_PATTERN_MAX_SZ = 15u;
+
+	// common data
+		// private and setters/getters instead?
+	float4 color;
+	float screenSpaceLineWidth;
+	float worldSpaceLineWidth;
+
+	void prepareGPUStipplePatternData(const nbl::core::vector<float>& stipplePatternCPURepresentation)
```

Comment:
1. name this `setStipplePatternData` and instead of a vector, get `core::SRange`
2. in `drawPolyline` get a CPULineStyle instead a gpu form LineStyle, and convert to gpu just before doing comparisons and copying it to mem (see `addLineStyle_Internal`)

Reply:
done